### ISSUE TITLE
<webview> with initially hidden visibility doesn't render

### ISSUE THREAD

**Issue by poiru** [electron/electron#8505]
In the repro case below, the webview is initially created with `visibility: hidden`. After a short delay, we switch to `visibility: visible`. The expected behaviour is that github.com is displayed in the webview regardless of the timeout duration.

Due to some regression in Electron 1.5.0, this results in the webview being permanently empty. Weirdly enough it works fine when using a short timeout (<20) on my macOS system.

### How to reproduce

**index.js**:
```js
'use strict';

const {app, BrowserWindow} = require('electron');
const path = require('path');

app.on('ready', () => {
  const win = new BrowserWindow({
    width: 500,
    height: 500,
    webPreferences: {
      nodeIntegration: true
    },
  });
  win.loadURL('file://' + path.join(__dirname, 'test.html'))
});
```

**test.html**:
```html
<!DOCTYPE html>
<html>
<style>
html, body, webview {
  position: fixed;
  margin: 0;
  width: 100%;
  height: 100%;
}

.hidden {
  visibility: hidden;
}
</style>
<body>
<webview id="webview" class="hidden" src="https://github.com"></webview>
</body>
<script>
setTimeout(() => {
  const webview = document.getElementById('webview');
  webview.style.visibility = 'visible';
}, 5000);
</script>
</html>
```

**Comment by kevinsawicki** [comment]
Yeah, can definitely reproduce this on 1.5.0, don't see it in Chrome 55 though, wonder if it still happens on #8501

**Comment by deepak1556** [comment]
Unable to reproduce on linux with the above example, might be platform specific. What does the `log-net-log` trace show ?

**Comment by poiru** [comment]
>Yeah, can definitely reproduce this on 1.5.0, don't see it in Chrome 55 though, wonder if it still happens on #8501

@kevinsawicki Unfortunately I can still reproduce with a build of the `chrome56` branch.

**Comment by poiru** [comment]
>Unable to reproduce on linux with the above example, might be platform specific. What does the `log-net-log` trace show ?

@deepak1556 It is roughly equivalent regardless of the timeout (i.e. github.com is definitely being loaded even if it is not rendered).

**Comment by pfrazee** [comment]
I'm experiencing this. I'm getting around it by `transform: translateX(-100%)` but that causes resizes to be quite slow, since every webview has to reflow (even if hidden).

**Comment by kevinsawicki** [comment]
> Unable to reproduce on linux with the above example, might be platform specific. 

Yeah, I'm only able to reproduce on macOS, not on Windows 10 or Ubuntu 16.04.

**Comment by poiru** [comment]
We worked around this by setting `visibility: hidden` after receiving the `did-finish-load` event for the webview. We use `z-index` to keep the webview invisible anyway and only use `visiblity: hidden` to throttle background/invisible tabs/webviews.

**Comment by brandonmp** [comment]
Adding to the pile since my exp. is a little different than those posted so far.

Using a `webview` via [a React wrapper](https://github.com/MarshallOfSound/react-electron-web-view)

As above, the problem only manifests when I'm testing on MacOS, not win/linux. 

I'm not using the `visibility` prop anywhere afaik (and not really many styles at all--just some flexbox layout stuff), so couldn't trace this to anything specific. Worse still was the fact that `webview` only failed to show up about 60% of the time (and only when I disable cache in dev tools). 

Eventually I realized there was probably some kind of race condition, and I narrowed it down (I hope) to my styles library, [Aphrodite](https://github.com/Khan/aphrodite), which injects styles at load time. 

I switched to regular react inline styles & now it seems to be working fine.

**Comment by kontrollanten** [comment]
I have a similar issue. I'm using the electron tabs module.

What doesn't works (nothing in webview is rendered):
**index.html**
```
<!doctype html>
<html>
	<head>
		<meta charset="utf-8">
		<title>Electron boilerplate</title>
		<link rel="stylesheet" href="index.css">
	</head>
	<body>
		<div class="flex-container">
			<div class="etabs-tabgroup">
				<div class="etabs-tabs"></div>
			        <div class="etabs-buttons"></div>
			</div>
			<div class="etabs-views"></div>
		</div>
	</body>
</html>
```

**index.css**
```
.flex-container {
	display: flex;
}
```

**electron tabs module** (which is called after DOMContentLoaded)
```
initWebview: function () {
        this.webview = document.createElement("webview");
        
        const tabWebviewDidFinishLoadHandler = function (e) {
            this.emit("webview-ready", this);
        };

        this.webview.addEventListener("did-finish-load", tabWebviewDidFinishLoadHandler.bind(this), false);
        
        this.webview.classList.add(this.tabGroup.options.viewClass);
        if (this.webviewAttributes) {
            let attrs = this.webviewAttributes;
            for (let key in attrs) {
                this.webview.setAttribute(key, attrs[key]);
            }
        }
        
        this.tabGroup.viewContainer.appendChild(this.webview);
    }
```
What actually works (the only change is that the style is inline):
**index.html**
```
<!doctype html>
<html>
	<head>
		<meta charset="utf-8">
		<title>Electron boilerplate</title>
		<style>
			.flex-container {
				display: flex;
			}
		</style>
	</head>
	<body>
		<div class="flex-container">
			<div class="etabs-tabgroup">
				<div class="etabs-tabs"></div>
			        <div class="etabs-buttons"></div>
			</div>
			<div class="etabs-views"></div>
		</div>
	</body>
</html>
```


Electron 1.6.6 
MacOS 10.12.4

Thank's @brandonmp for declaring your solution, it helped me out.

**Comment by lukeapage** [comment]
I appear to be getting this issue on windows. It is intermittant - but we have a webview which we hide when loading and the webview goes through a few redirects. On one of those redirected, not always the same one, it fails to load. A debugger on the webview gets disconnected, we get a destroyed event and we get a load-did-fail event

In the screenshot I've removed the URL's - there is nothing wrong with them

![image](https://cloud.githubusercontent.com/assets/309321/26672955/6aec90d2-46bb-11e7-80e4-ac2ee5712b58.png)

**Comment by CharlieHess** [comment]
At Slack, we employed the same workaround that @poiru describes, although a bit more painful because our `webview`s can be in various states of load – the full `slack.com` or a pared down "min" page for teams that are rarely active (this means we're occasionally navigating when the user switches teams).

Also worth nothing that even after this workaround, `document.visibilityState` on the embedded page cannot be trusted: it will always return `'visible'`. This causes further pain for us because `slack.com` relies on `visibilitychange` for a few things and now we have to override those with our _own_ "is the page really visible?" check.

In a nutshell, this bug really hurts. 🤕

**Comment by poiru** [comment]
@CharlieHess Re: `document.visibilityChange`, see #9178. (I had to revert the fix for `<webview>` though.)

**Comment by CharlieHess** [comment]
@poiru nice work! We're stuck on that `webview` life though. 😔

**Comment by CharlieHess** [comment]
I made a minimal repro for the issues we're seeing in the Slack app. To summarize:

* Relying on `z-index` for visibility can cause layered `webview`s to bleed into each other
* If you switch `webview`s at just the right time, you can _still_ get white screens

Repo is up at https://github.com/CharlieHess/webview-stack.

**Comment by CharlieHess** [comment]
Another repro case that breaks this fix is when the embedded page decides to navigate itself. In that case the `did-start-loading` event doesn't seem to fire early enough (also tried `did-get-redirect-request`, `did-navigate`, etc. to no avail).

To see it in a minimal repro:

1. Checkout https://github.com/CharlieHess/webview-stack/tree/slack-teams
1. Edit `webViews` in `App.jsx` with some Slack teams that you can login to
1. Login to at least two of the teams
1. Turn on 2FA or otherwise invalidate the session on the _unselected_ team (navigate to https://your-team.slack.com/account/settings/2fa_choose in the browser)
1. Switch to that team, observe a blanked out `webview`

**Comment by guillaumearm** [comment]
Is this bug is still present in `electron@2.0.2` ?

**Comment by Polunom** [comment]
@guillaumearm yep, also broken on `electron@3.0.0-beta.12`

**Comment by Polunom** [comment]
@guillaumearm Yeah, it seems to be broken for all the `electron@2.x.x` versions I've tried. Interestingly, it seems to be fixed on the the newest beta, `electron@3.0.0-beta.12`.

Screenshot using @poiru's test code:
![It works!](https://i.imgur.com/DzUNN9I.png)

**Comment by codebytere** [comment]
Given that `3.0.0` is now stable, and this issue appears to be fixed there, i'm going to go ahead and close this out.
